Modules/_ctypes/stgdict.c

```diff
@@ -16,201 +16,121 @@
 #endif
 #include "ctypes.h"
 
-/******************************************************************/
-/*
-  StdDict - a dictionary subclass, containing additional C accessible fields
-
-  XXX blabla more
-*/
-
-/* Seems we need this, otherwise we get problems when calling
- * PyDict_SetItem() (ma_lookup is NULL)
+/* This file relates to StgInfo -- type-specific information for ctypes.
+ * See ctypes.h for details.
  */
+
+// Get a PyCTypeDataObject. These return -1 on error, 0 if "not found", 1 on OK.
 static int
-PyCStgDict_init(StgDictObject *self, PyObject *args, PyObject *kwds)
+_stginfo_from_type(ctypes_state *state, PyTypeObject *type, StgInfo **result)
 {
-    if (PyDict_Type.tp_init((PyObject *)self, args, kwds) < 0)
-        return -1;
-    self->format = NULL;
-    self->ndim = 0;
-    self->shape = NULL;
-    return 0;
+    *result = NULL;
+    if (!PyObject_IsInstance((PyObject *)type, (PyObject *)state->PyCType_Type)) {
+        // not a ctypes class.
+        return 0;
+    }
+    StgInfo *info = PyObject_GetTypeData((PyObject *)type, state->PyCType_Type);
```

Comment:
According to the docs, `PyObject_GetTypeData` raises an exception if `NULL` is returned, however we know that the implementation (currently) does not raise an exception. It may be nice to future readers of the code to make this more explicit, for example by adding an assert.

```suggestion
    StgInfo *info = PyObject_GetTypeData((PyObject *)type, state->PyCType_Type);
    assert(info != NULL);
```